include/nbl/builtin/hlsl/emulated/float64_t_impl.hlsl

```diff
@@ -0,0 +1,312 @@
+
+template<bool FlushDenormToZero>
+inline uint64_t castFloat32ToStorageType(float32_t val)
+{
+    if (FlushDenormToZero)
+    {
+        const uint64_t sign = uint64_t(ieee754::extractSign(val)) << 63;
+        if (tgmath::isInf(val))
+            return ieee754::traits<float64_t>::inf | sign;
+        uint32_t asUint = ieee754::impl::bitCastToUintType(val);
+        const int f32BiasedExp = int(ieee754::extractBiasedExponent(val));
+        if (f32BiasedExp == 0)
+            return sign;
+        const uint64_t biasedExp = uint64_t(f32BiasedExp - ieee754::traits<float32_t>::exponentBias + ieee754::traits<float64_t>::exponentBias) << (ieee754::traits<float64_t>::mantissaBitCnt);
+        const uint64_t mantissa = (uint64_t(ieee754::traits<float32_t>::mantissaMask) & asUint) << (ieee754::traits<float64_t>::mantissaBitCnt - ieee754::traits<float32_t>::mantissaBitCnt);
+
+        return sign | biasedExp | mantissa;
+    }
+    else
+    {
+        // static_assert(false);
+        return 0xdeadbeefbadcaffeull;
+    }
+};
+
+NBL_CONSTEXPR_INLINE_FUNC bool isZero(uint64_t val)
+{
+    return (val << 1) == 0;
+}
+
+// TODO: where do i move this function? also rename
+template <typename Int>
+inline int _findMSB(Int val)
+{
+    //static_assert(is_integral<Int>::value);
+#ifndef __HLSL_VERSION
+    return nbl::hlsl::findMSB(val);
+#else
+    return firstbithigh(val);
+#endif
+}
+
+template <>
+inline int _findMSB(uint64_t val)
+{
+#ifndef __HLSL_VERSION
+    return nbl::hlsl::findMSB(val);
+#else
+    int msbHigh = firstbithigh(uint32_t(val >> 32));
+    int msbLow = firstbithigh(uint32_t(val));
+    return msbHigh != -1 ? msbHigh + 32 : msbLow;
+#endif
+}
+
+inline uint64_t castToUint64WithFloat64BitPattern(uint64_t val)
+{
+    if (isZero(val))
+        return val;
+
+    int exp = _findMSB(val);
+    uint64_t mantissa;
+
+    int shiftCnt = 52 - exp;
+    if (shiftCnt >= 0)
+    {
+        mantissa = val << shiftCnt;
+    }
+    else
+    {
+        const int shiftCntAbs = -shiftCnt;
+        uint64_t roundingBit = 1ull << (shiftCnt - 1);
+        uint64_t stickyBitMask = roundingBit - 1;
+        uint64_t stickyBit = val & stickyBitMask;
+
+        mantissa = val >> shiftCntAbs;
+
+        if ((val & roundingBit) && (!stickyBit))
+        {
+            bool isEven = mantissa & 1;
+            if (!isEven)
+                mantissa++;
+        }
+        else if ((val & roundingBit) && (stickyBit || (mantissa & 1)))
+            val += roundingBit;
+
+        //val += (1ull << (shiftCnt)) - 1;
+        //mantissa = val >> shiftCntAbs;
+
+        if (mantissa & 1ull << 53)
+        {
+            mantissa >>= 1;
+            exp++;
+        }
+    }
+    mantissa &= ieee754::traits<float64_t>::mantissaMask;
+    const uint64_t biasedExp = uint64_t(ieee754::traits<float64_t>::exponentBias + exp) << ieee754::traits<float64_t>::mantissaBitCnt;
+
+    return biasedExp | mantissa;
+};
+
+inline uint64_t castToUint64WithFloat64BitPattern(int64_t val)
```

Comment:
I really dislike the namings of these functions, why not just move them into their call site?